src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\AddressingBundle\EventListener;
+
+use Sylius\Component\Addressing\Checker\CountryProvincesDeletionCheckerInterface;
+use Sylius\Component\Addressing\Checker\ZoneDeletionCheckerInterface;
+use Sylius\Component\Addressing\Model\CountryInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Webmozart\Assert\Assert;
+
+final class ZoneMemberIntegrityListener
+{
+    public function __construct(
+        private SessionInterface $session,
+        private ZoneDeletionCheckerInterface $zoneDeletionChecker,
+        private CountryProvincesDeletionCheckerInterface $countryProvincesDeletionChecker
+    ) {
+    }
+
+    public function protectFromRemovingZone(GenericEvent $event): void
+    {
+        $zone = $event->getSubject();
+        Assert::isInstanceOf($zone, ZoneInterface::class);
+
+        if (!$this->zoneDeletionChecker->isDeletable($zone)) {
+            /** @var FlashBagInterface $flashes */
+            $flashes = $this->session->getBag('flashes');
+            $flashes->add('error', [
+                'message' => 'sylius.resource.delete_error',
+                'parameters' => ['%resource%' => 'zone'],
+            ]);
+
+            $event->stopPropagation();
+        }
+    }
+
+    public function protectFromRemovingProvinceWithinCountry(GenericEvent $event): void
+    {
+        /** @var CountryInterface $country */
+        $country = $event->getSubject();
+        Assert::isInstanceOf($country, CountryInterface::class);
```

Comment:
Shouldn't we check to remove the subject that is `ProvinceInterface`?

Reply:
The problem is we don't delete the Province directly but `orphanRemoval` configured in relation within a Country does it for us. That makes me unable to listen for an event `sylius.province.pre_delete` and I had to use `sylius.country.pre_update`.